Refactor/update margin format

LurN opened the pull request:
## Feat
- Change margin format from `x:y` to `[x, y]`, where either or both brackets can be replaced by parenthesis
() : exclusive
[] : inclusive

yulric commented on R/utils.R:
```diff
@@ -6,7 +6,14 @@
 #' @examples
 get_margins <- function (chars) {
-  return (trimws(strsplit(chars, ":")[[1]]))
+  chars <- trimws(chars)
```
Don't mutate variables.

yulric commented on R/utils.R:
```diff
@@ -6,7 +6,14 @@
 get_margins <- function (chars) {
-  return (trimws(strsplit(chars, ":")[[1]]))
+  chars <- trimws(chars)
+  left_is_exclusive <- substr(chars, 1, 1) == "("
```
We should use interval terminology which can be reference here https://en.wikipedia.org/wiki/Interval_(mathematics).
For e.g., this variables would be called `is_left_open`.

yulric commented on R/utils.R:
```diff
@@ -28,7 +35,7 @@ is_numeric <- function(chars) {
 #' @examples
 is_rec_from_range <- function(var_details_row) {
-  return (grepl(":", var_details_row$recFrom, fixed=TRUE))
+  return (grepl(",", var_details_row$recFrom, fixed = TRUE))
```
I think it would be better to use the `get_margins` method for this. If the margins returned are the same then this method would return TRUE otherwise it would return FALSE.
The `get_margins` method would have to be updated to handle recFrom values that are not intervals.

yulric commented on R/utils.R:
```diff
@@ -6,7 +6,14 @@
+  right_is_exclusive <- substr(chars, nchar(chars), nchar(chars) + 1) == ")"
+
+  margins <- strsplit(substr(chars, 2, nchar(chars) - 1), ",")[[1]]
+  if (left_is_exclusive) margins[1] = as.character(as.numeric(margins[1]) + 1);
```
Remember an open interval does not imply that the margins needs to be reduced by 1. For e.g., a recFrom value of (1,2] means that the effective range us 0.99999....,2 whereas in your case it would be 0 to 2 which is wrong.
Same issue with the next line of code.

yulric reviewed (changes requested):
Can you update the integration tests to have the following test cases?
1. A case for [ ]
2. A case for ( )
3. A case for [ )
4. A case for ( ]

LurN commented:
@yulric I've updated it with those changes

yulric reviewed (changes requested):
I updated the tests to make sure we're covering all the cases for intervals and I found 2 errors:
1. The generated Values for the ALCA_5A1 variable is wrong. It should be 3 intervals and not a lot of Value nodes.
2. The transformation for ALW_2A1 is incorrect. The if nodes look right but I don't think its setting the right values if they evaluate to true.
I'll attach

The attached TXT file below has what the code is outputting (Github does not allow you to attach XML files hence the text file)
[expected-pmml.txt](https://github.com/Big-Life-Lab/recodeflow/files/5876852/expected-pmml.txt)

LurN commented:
> I updated the tests to make sure we're covering all the cases for intervals and I found 2 errors:
> 
> 1. The generated Values for the ALCA_5A1 variable is wrong. It should be 3 intervals and not a lot of Value nodes.
> 2. The transformation for ALW_2A1 is incorrect. The if nodes look right but I don't think its setting the right values if they evaluate to true.
>    I'll attach
> 
> The attached TXT file below has what the code is outputting (Github does not allow you to attach XML files hence the text file)
> [expected-pmml.txt](https://github.com/Big-Life-Lab/recodeflow/files/5876852/expected-pmml.txt)

@yulric 
1. For ALCA_5A1, should it have 5 Interval nodes? It looks like there's 4 valid value ranges, then the missing range.
2. The format of the DerivedField section looks different for cont vs cat, where cont doesn't have Constant nodes directly within the if nodes, should they also have those nodes? And is that the issue ur talking about?

Here's the latest output: [output.txt](https://github.com/Big-Life-Lab/recodeflow/files/5908467/output.txt)

Also, I've updated the `closure` property on the Interval nodes to match the types in `recFrom` to use `closedClosed`, `openClosed` etc

yulric commented:
> > I updated the tests to make sure we're covering all the cases for intervals and I found 2 errors:
> > 
> > 1. The generated Values for the ALCA_5A1 variable is wrong. It should be 3 intervals and not a lot of Value nodes.
> > 2. The transformation for ALW_2A1 is incorrect. The if nodes look right but I don't think its setting the right values if they evaluate to true.
> >    I'll attach
> > 
> > The attached TXT file below has what the code is outputting (Github does not allow you to attach XML files hence the text file)
> > [expected-pmml.txt](https://github.com/Big-Life-Lab/recodeflow/files/5876852/expected-pmml.txt)
> 
> @yulric
> 
> 1. For ALCA_5A1, should it have 5 Interval nodes? It looks like there's 4 valid value ranges, then the missing range.
> 2. The format of the DerivedField section looks different for cont vs cat, where cont doesn't have Constant nodes directly within the if nodes, should they also have those nodes? And is that the issue ur talking about?
> 
> Here's the latest output: [output.txt](https://github.com/Big-Life-Lab/recodeflow/files/5908467/output.txt)
> 
> Also, I've updated the `closure` property on the Interval nodes to match the types in `recFrom` to use `closedClosed`, `openClosed` etc
 @LurN 

1. RACA_6A looks good
2. For ALCA_5A1
     1.  The last interval (997 to 999) should be represented as Value nodes. Right now, all invalid values should be represented as Value nodes
     2. Can you add the property field to the Interval nodes. So this would be a property value of valid
3. Sorry about this Luke, but for the parts where you're setting values to `<Constant missing="true"/>`, can you instead set them to their recTo values with dataType string? For eg., `<Constant dataType="string">NA::a</Constant>`?
4. The transformations for ALW_2A1 are still wrong? Right now it's not setting it to the correct value if a condition is met. The XML right now in JS is,
```
if(ALW_2A1 >= 1 && ALW_2A1 <= 10) {

} ...etc
```
Its missing what should go in the brackets
5. For the transformed variables. `ADL_01, ALW_2A1, DHHGAGE_cont`, can you add the Value nodes for the missing and invalid recTo columns? So the NA::a and NA::b?
6. Is it possible to fix this `don&apos;t` so it shows the `'` instead of `&apos;`?

LurN commented:
> 1. For ALCA_5A1
>    
>    1. The last interval (997 to 999) should be represented as Value nodes. Right now, all invalid values should be represented as Value nodes
>    2. Can you add the property field to the Interval nodes. So this would be a property value of valid
> 2. Sorry about this Luke, but for the parts where you're setting values to `<Constant missing="true"/>`, can you instead set them to their recTo values with dataType string? For eg., `<Constant dataType="string">NA::a</Constant>`?
> 3. The transformations for ALW_2A1 are still wrong? Right now it's not setting it to the correct value if a condition is met. The XML right now in JS is,
> 
> ```
> if(ALW_2A1 >= 1 && ALW_2A1 <= 10) {
> 
> } ...etc
> ```
> 
> Its missing what should go in the brackets
> 4. For the transformed variables. `ADL_01, ALW_2A1, DHHGAGE_cont`, can you add the Value nodes for the missing and invalid recTo columns? So the NA::a and NA::b?
> 5. Is it possible to fix this `don&apos;t` so it shows the `'` instead of `&apos;`?

@yulric 

1. 🚀 
2. 👍 
3. Ok, I think i've fixed this, I'm not certain the structure is correct
4. Sounds good, I've done that. Do I need to remove duplicates? For example, there will be multiple `NA::b` nodes. And is this the right format, just the same as the others?:
```
<Value value="NA::b" displayValue="missing">
  <Extension name="catLabelLong" value="missing"/>
</Value>
```
5. I don't see any options in the XML package to prevent it from replacing the special characters, but I can try doing something like recursively go through nodes, their values and attributes, and

yulric commented:
> > 1. For ALCA_5A1
> >    
> >    1. The last interval (997 to 999) should be represented as Value nodes. Right now, all invalid values should be represented as Value nodes
> >    2. Can you add the property field to the Interval nodes. So this would be a property value of valid
> > 2. Sorry about this Luke, but for the parts where you're setting values to `<Constant missing="true"/>`, can you instead set them to their recTo values with dataType string? For eg., `<Constant dataType="string">NA::a</Constant>`?
> > 3. The transformations for ALW_2A1 are still wrong? Right now it's not setting it to the correct value if a condition is met. The XML right now in JS is,
> > 
> > ```
> > if(ALW_2A1 >= 1 && ALW_2A1 <= 10) {
> > 
> > } ...etc
> > ```
> > 
> > 
> > Its missing what should go in the brackets
> > 4. For the transformed variables. `ADL_01, ALW_2A1, DHHGAGE_cont`, can you add the Value nodes for the missing and invalid recTo columns? So the NA::a and NA::b?
> > 5. Is it possible to fix this `don&apos;t` so it shows the `'` instead of `&apos;`?
> 
> @yulric
> 
> 1. 🚀
> 2. 👍
> 3. Ok, I think i've fixed this, I'm not certain the structure is correct
> 4. Sounds good, I've done that. Do I need to remove duplicates? For example, there will be multiple `NA::b` nodes. And is this the right format, just the same as the others?:
> 
> ```
> <Value value="NA::b" displayValue="missing">
>   <Extension name="catLabelLong" value="missing"/>
> </Value>
> ```
> 
> 1. I don't see any options in the XML package to prevent it from replacing the special characters, but I can try doing something like recursively go through nodes, their values and attributes, and

@LurN 

I updated the variable details test file so that it has fewer rows for DHHGAGE_cont. This will make it easier to debug issues.

Everything looks good except for the transformations for the `DHHGAGE_cont` variable. If a condition is met its not setting it to the `recTo` value but to 1 and then 2 and then 3 etc.

For point 4, yes, can you remove the duplicate NA::b rows for the `DerivedField`?

LurN commented:
@yulric I've made the changes, and I've updated the expected-pmml file with the results

yulric reviewed (approved):
Looks good. If you've updated the PMML file then ready to merge.
